src/librustc_typeck/check/mod.rs

```diff
@@ -3557,9 +3557,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             _ => self.warn_if_unreachable(expr.id, expr.span, "expression")
         }
 
-        // Any expression that produces a value of type `!` must have diverged
-        if ty.is_never() {
-            self.diverges.set(self.diverges.get() | Diverges::Always);
+        // Any expression that produces a value of an uninhabited type must have diverged.
+        if ty.conservative_is_uninhabited() {
+            let always = if ty.is_never() {
+                Diverges::Always
+            } else {
+                // We avoid linting in this case for
+                // consistency with previous versions.
+                Diverges::UnwarnedAlways
```

Comment:
I have a question here -- if we later process an expression that produces a value of type `!`, would we lint then? e.g., `foo(a, b)` where `a: &!` and `b: !`.

Reply:
I seem to recall reading that `UnwarnedAlways` can never be converted to `WarnedAlways`... so I am guessing not? Or maybe I am misinterpreting/misremembering how that enum works.

Reply:
Hmm, yes, I suppose it wouldn't lint a later unreachable expression, which probably isn't what we want. I'll check that.

Reply:
Actually, this shouldn't be an issue, because the only time we use `Diverges::UnwarnedAlways` is with an empty function that can't be executed, but we don't want to trigger the lint for (as in https://github.com/rust-lang/rust/pull/50262#pullrequestreview-116439632). In such a case, no more expressions are available to trigger a divergence warning.